# HeroesOfPymoli In-game Purchase Analysis
Below is an analysis of in-game purchases by players of HeroesOfPymoli. You will find a breakdown of players and purchases by both gender and age along with a look at the top players and items available for purchase. Included at the bottom are three trends observed through the analysis provided.

In [1]:
# Dependencies and Setup
import pandas as pd

# File to Load (Remember to Change These)
file_to_load = "Resources/purchase_data.csv"

# Read Purchasing File and store into Pandas data frame, display sample rows to see what data is available
purchase_data = pd.read_csv(file_to_load)
purchase_data.head(5)

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

## Player Count

* Identify and display the total number of unique players

In [2]:
unique_player_count = len(purchase_data["SN"].unique())

total_players_df = pd.DataFrame({"Total Players":[unique_player_count]})
total_players_df

Total Players
0            576

## Purchasing Analysis (Total)

* Calculate the number of unique items, total number of purchases, average purchase price and total revenue.
    - calculate necessary values
    - create and display summary dataframe, formatted for readability
   

In [3]:
#perform calculations on dataframe
unique_items_count = len(purchase_data["Item ID"].unique())
total_purchases = len(purchase_data["Item ID"])
ave_purchase_price = purchase_data["Price"].mean()
total_revenue = purchase_data["Price"].sum()

#create, display and format datafram
purchasing_analysis_df = pd.DataFrame({"Number of Unique Items":[unique_items_count],
                                    "Total Number of Purchases":[total_purchases],
                                   "Average Purchase Price":[ave_purchase_price],
                                   "Total Revenue":[total_revenue]}).round(decimals=2)
purchasing_analysis_df["Average Purchase Price"] = purchasing_analysis_df["Average Purchase Price"].astype(float).map("${:,.2f}".format)
purchasing_analysis_df["Total Revenue"] = purchasing_analysis_df["Total Revenue"].astype(float).map("${:,.2f}".format)
purchasing_analysis_df

Number of Unique Items  Total Number of Purchases Average Purchase Price  \
0                     179                        780                  $3.05   

  Total Revenue  
0     $2,379.77

## Gender Demographics

* Calculate and dispay the percentage and unique count of male, female and other/non-disclosed players
    - create new dataframe from original loaded file, limited to unique users
    - count users by gender and calculate percentage using unique player count found in "Player Count" section
    - store this calculated information in a new dataframe, formatted for readability

In [4]:
#create unique player dataframe
reduced_purchase_data = purchase_data[["SN","Gender","Age"]]
reduced_purchase_data = reduced_purchase_data.drop_duplicates()

#count users and calculate percent
gender_count = reduced_purchase_data["Gender"].value_counts()
gender_percent = (gender_count/unique_player_count)*100

#create, display and format summary dataframe
gender_demographics = pd.DataFrame({"Total Count":gender_count,
                                   "Percentage of Players":gender_percent})
gender_demographics["Percentage of Players"] = gender_demographics["Percentage of Players"].astype(float).map("{:.2f}%".format)
gender_demographics

Total Count Percentage of Players
Male                           484                84.03%
Female                          81                14.06%
Other / Non-Disclosed           11                 1.91%

## Purchasing Analysis (Gender)

* By gender, calculate and display purchase count, average purchase price, total purchase price and average purchase total per person
    - create new dataframe from original loaded file (this is not limited to unique users as it is looking at all purchases)
    - calculate and store necessary variables
    - create and display new summary dataframe, formatted for readability

In [5]:
#create new dataframe
purchasing_analysis = purchase_data[["Purchase ID", "Price", "Gender", "SN"]]

#count purchases and calculate values/costs
purchase_count = purchasing_analysis.groupby("Gender").count()["SN"]
total_purchase_value = purchasing_analysis.groupby("Gender").sum()["Price"]
ave_purchase_cost = purchasing_analysis.groupby("Gender").mean()["Price"]
ave_purchase_per_person = total_purchase_value / gender_demographics["Total Count"]

#create, display and format summary dataframe
purchasing_analysis_summary_df = pd.DataFrame({"Purchase Count":purchase_count,
                                              "Average Purchase Price":ave_purchase_cost,
                                              "Total Purchase Value":total_purchase_value,
                                              "Avg Total Purchase Per Person":ave_purchase_per_person})

purchasing_analysis_summary_df["Average Purchase Price"] = purchasing_analysis_summary_df["Average Purchase Price"].astype(float).map("${:.2f}".format)
purchasing_analysis_summary_df["Total Purchase Value"] = purchasing_analysis_summary_df["Total Purchase Value"].astype(float).map("${:,.2f}".format)
purchasing_analysis_summary_df["Avg Total Purchase Per Person"] = purchasing_analysis_summary_df["Avg Total Purchase Per Person"].astype(float).map("${:.2f}".format)

purchasing_analysis_summary_df

Purchase Count Average Purchase Price  \
Gender                                                         
Female                            113                  $3.20   
Male                              652                  $3.02   
Other / Non-Disclosed              15                  $3.35   

                      Total Purchase Value Avg Total Purchase Per Person  
Gender                                                                    
Female                             $361.94                         $4.47  
Male                             $1,967.64                         $4.07  
Other / Non-Disclosed               $50.19                         $4.56

## Age Demographics

* Summarize the unique player counts by age ranges
    - create new dataframe holding screenname and age from unique players dataframe create in the Gender Demographics section
    - create bins for age ranges and assign names for the ranges
    - apply these bins to the dataframe
    - group by the age ranges and provide a count of unique players that fall in that range
    - display dataframe, formatted for readability

In [6]:
#create new dataframe
age_demographics_df = reduced_purchase_data[["SN", "Age"]]
#age_demographics_df.head()

#create, label and apply bins
bins = [0, 9.9, 14.9, 19.9, 24.9, 29.9, 34.9, 39.9, 50]
bin_names = ["<10", "10-14", "15-19", "20-24", "25-29", "30-34", "35-39", "40+"]
age_demographics_df["Age Range"] = pd.cut(age_demographics_df["Age"], bins, labels=bin_names, include_lowest=True)
#age_demographics_df.head()

#group dataframe, format and display
grouped_ages = age_demographics_df.groupby("Age Range").count()["SN"]
age_percent = (grouped_ages / unique_player_count)*100
grouped_ages_df = pd.DataFrame({"Unique Player Count":grouped_ages,
                               "Percentage of Players":age_percent})
grouped_ages_df["Percentage of Players"] = grouped_ages_df["Percentage of Players"].astype(float).map("{:.2f}%".format)
grouped_ages_df

Unique Player Count Percentage of Players
Age Range                                           
<10                         17                 2.95%
10-14                       22                 3.82%
15-19                      107                18.58%
20-24                      258                44.79%
25-29                       77                13.37%
30-34                       52                 9.03%
35-39                       31                 5.38%
40+                         12                 2.08%

## Purchasing Analysis (Age)

* Summarize purchases by age range based on full original loaded data set
    - create a copy of the original loaded dataframe for data integrity
    - apply previously created age ranges to the dataframe
    - perform calculations on the dataframe to gather (by age group) purchase count, average purchase price, total purchase value and average purchase total per person
    - create and display summary dataframe, formatted for readability

In [7]:
# create new dataframe and apply age ranges
binned_purchase_data = purchase_data
binned_purchase_data["Age Range"] = pd.cut(binned_purchase_data["Age"], bins, labels=bin_names, include_lowest=True)

# perform and store calculations
binned_purchase_count = binned_purchase_data.groupby("Age Range").count()["Purchase ID"]
binned_total_purchase_value = binned_purchase_data.groupby("Age Range").sum()["Price"]
binned_ave_price = binned_total_purchase_value / binned_purchase_count
binned_avg_purchase_per_person = binned_total_purchase_value / grouped_ages

#create, display and format summary dataframe
binned_purchasing_analysis_df = pd.DataFrame({"Puchase Count":binned_purchase_count,
                                             "Average Purchase Price":binned_ave_price,
                                             "Total Purchase Value":binned_total_purchase_value,
                                             "Avg Total Purchase per Person": binned_avg_purchase_per_person})
binned_purchasing_analysis_df["Average Purchase Price"] = binned_purchasing_analysis_df["Average Purchase Price"].astype(float).map("${:,.2f}".format)
binned_purchasing_analysis_df["Total Purchase Value"] = binned_purchasing_analysis_df["Total Purchase Value"].astype(float).map("${:,.2f}".format)
binned_purchasing_analysis_df["Avg Total Purchase per Person"] = binned_purchasing_analysis_df["Avg Total Purchase per Person"].astype(float).map("${:,.2f}".format)

binned_purchasing_analysis_df

Puchase Count Average Purchase Price Total Purchase Value  \
Age Range                                                              
<10                   23                  $3.35               $77.13   
10-14                 28                  $2.96               $82.78   
15-19                136                  $3.04              $412.89   
20-24                365                  $3.05            $1,114.06   
25-29                101                  $2.90              $293.00   
30-34                 73                  $2.93              $214.00   
35-39                 41                  $3.60              $147.67   
40+                   13                  $2.94               $38.24   

          Avg Total Purchase per Person  
Age Range                                
<10                               $4.54  
10-14                             $3.76  
15-19                             $3.86  
20-24                             $4.32  
25-29                             $3.81  
30-34                             $4.12  
35-39                             $4.76  
40+                               $3.19

## Top Spenders

* Identify the top 5 spenders in the game by total purchase value
    - calculate total spending, amount of purchases and the average spending by unique screen name
    - create summary dataframe, sort descending by total purchase value and format for readability
    - display top 5 rows of summary dataframe

In [8]:
#using original loaded file, calculate necessary values
spending_by_sn = purchase_data.groupby("SN").sum()["Price"]
purchases_by_sn = purchase_data.groupby("SN").count()["Purchase ID"]
ave_spent_by_sn = purchase_data.groupby("SN").mean()["Price"]

#create, display and format summary dataframe
top_spenders_df = pd.DataFrame({"Purchase Count":purchases_by_sn,
                               "Average Purchase Price":ave_spent_by_sn,
                               "Total Purchase Value":spending_by_sn}).sort_values(["Total Purchase Value"],ascending=False)
top_spenders_df["Average Purchase Price"] = top_spenders_df["Average Purchase Price"].astype(float).map("${:,.2f}".format)
top_spenders_df["Total Purchase Value"] = top_spenders_df["Total Purchase Value"].astype(float).map("${:,.2f}".format)
top_spenders_df.head(5)

Purchase Count Average Purchase Price Total Purchase Value
SN                                                                     
Lisosia93                 5                  $3.79               $18.96
Idastidru52               4                  $3.86               $15.45
Chamjask73                3                  $4.61               $13.83
Iral74                    4                  $3.40               $13.62
Iskadarya95               3                  $4.37               $13.10

## Most Popular Items

* Identify the 5 most popular items in the game by purchase count
    - create new dataframe from original loaded file to contain a full list of purchases
    - group dataframe by item ID and name
    - calculate total purchases of each item, average item price and total purchase value
    - create summary dataframe and format for readability

In [9]:
#create new dataframe
most_popular_df = purchase_data[["Item ID", "Item Name", "Price"]]

#group dataframe
grp_most_popular_df = most_popular_df.groupby(["Item ID","Item Name"])

#calculate purchase data
pop_purchase_count = grp_most_popular_df["Item ID"].count()
pop_ave_item_price = grp_most_popular_df["Price"].mean()
pop_total_purchase_value = grp_most_popular_df["Price"].sum()

#create summary dataframe, sort descending by purchase count
comb_most_pop_df = pd.DataFrame({"Purchase Count": pop_purchase_count,
                                "Item Price":pop_ave_item_price,
                                "Total Purchase Value":pop_total_purchase_value}).sort_values(["Purchase Count"],ascending=False)

#create formatted dataframe for readability and to retain usable datatypes for next section
comb_most_pop_format = comb_most_pop_df.head(5).style.format({"Item Price":"${:,.2f}","Total Purchase Value":"${:,.2f}"})
comb_most_pop_format

## Most Profitable Items

* Identify the 5 most profitable items by total purchase value
    - use unformatted dataframe from previous section to resort data based on total purchase value
    - create summary dataframe and format for readability

In [10]:
#resort unformatted dataframe from previous section
most_profitable_items = comb_most_pop_df.sort_values(["Total Purchase Value"], ascending=False)

#create and display formatted dataframe
most_profitable_items_format = most_profitable_items.head(5).style.format({"Item Price":"${:,.2f}","Total Purchase Value":"${:,.2f}"})
most_profitable_items_format

----------------------------------------------------------------------------------------------------------------------------

### Three Noteable Trends
1. 84% of unique players identify as male. This same proportion carries over to the total purchases, in that 84% of all purchases are made by players who identify as male.
2. 44% of unique players are between the ages of 20-24 with a similar percentage, 47%, of all purchases being made by the same age group.
3. The top two most popular in-game purchasable items, Final Critic and Oathbreaker, Last Hope fo the Breaking Storm are also the top two most profitable items in the same order.